modin/pandas/test/dataframe/test_protocol.py

```diff
@@ -0,0 +1,276 @@
+
+    df2 = df.__dataframe__()
+
+    for col_name in df.columns:
+        #assert convert_column_to_array(df2.get_column_by_name(col_name) == df[col_name].tolist()
+        assert df2.get_column_by_name(col_name).null_count == 2
+        assert df[col_name].dtype == df2[col_name].dtype
+
+
+def test_missing_from_masked():
+    df = pd.DataFrame(
+        {
+            "x": np.array([1, 2, 3, 4, 0]),
+            "y": np.array([1.5, 2.5, 3.5, 4.5, 0]),
+            "z": np.array([True, False, True, True, True]),
+        }
+    )
+
+    df2 = df.__dataframe__()
+
+    for col_name in df.columns:
+        #assert convert_column_to_array(df2.get_column_by_name(col_name) == df[col_name].tolist()
+        assert df[col_name].dtype == df2[col_name].dtype
+    
+    dict_null = {}
+    for col_name in df.columns:
+        num_null = np.random.randint(5)
+        list_null = []
+        num_nulls = 0
+        for ind in num_null:
+            ind_null = np.random.randint(5)
+            if ind_null not in list_null:
+                (df[col_name])[ind_null] = None
+                list_null.append(ind_null)
+                num_nulls += 1
+        dict_null[col_name] = num_nulls
+    
+    df2 = df.__dataframe__()
+
+    assert df2.get_column_by_name("x").null_count == dict_null["x"]
+    assert df2.get_column_by_name("y").null_count == dict_null["y"]
+    assert df2.get_column_by_name("z").null_count == dict_null["z"]
+
+
+def test_string():
+    df = pd.DataFrame({"A": ["a", None, "cdef", "", "g"]})
+    col = df.__dataframe__().get_column_by_name("A")
+
+    assert col.size == 5
+    assert col.null_count == 1
+    assert col.dtype[0] == DTypeKind.STRING
+    if StorageFormat.get() != "Omnisci":
+        assert col.describe_null == (4, 0)
+
+    df2 = df.__dataframe__()
+    assert df2.A.tolist() == df.A.tolist()
+    assert df2.get_column_by_name("A").null_count == 1
+    if StorageFormat.get() != "Omnisci":
+        assert df2.get_column_by_name("A").describe_null == (4, 0)
+    assert df2.get_column_by_name("A").dtype[0] == DTypeKind.STRING
+
+    df_sliced = df[1:]
+    col = df_sliced.__dataframe__().get_column_by_name("A")
+    assert col.size == 4
+    assert col.null_count == 1
+    assert col.dtype[0] == DTypeKind.STRING
+    if StorageFormat.get() != "Omnisci":
+        assert col.describe_null == (4, 0)
+
+    df2 = df_sliced.__dataframe__()
+    assert df2.A.tolist() == df_sliced.A.tolist()
+    assert df2.get_column_by_name("A").null_count == 1
+    if StorageFormat.get() != "Omnisci":
+        assert df2.get_column_by_name("A").describe_null == (4, 0)
+    assert df2.get_column_by_name("A").dtype[0] == DTypeKind.STRING
+
+
+def test_object():
+    df = pd.DataFrame({"x": np.array([None, True, False])})
+    col = df.__dataframe__().get_column_by_name("x")
+
+    assert col.size == 3
+
+    with pytest.raises(NotImplementedError):
+        assert col.dtype
+    with pytest.raises(NotImplementedError):
+        assert col.describe_null
```

Comment:
the same question, why does it raise an exception (`ValueError` that was here before or `NotImplementedError`)?

Reply:
Given [link1](https://data-apis.org/dataframe-protocol/latest/design_requirements.html#:~:text=Object%20dtype%20does%20not%20need%20to%20be%20supported) and [link2](https://data-apis.org/dataframe-protocol/latest/purpose_and_scope.html#:~:text=Rationale%3A%20object%20dtypes%20are%20inefficient%20and%20may%20contain%20anything%20(so%20hard%20to%20support%20in%20a%20sensible%20way)%3B%20nested/structures%20dtypes%20may%20be%20supported%20in%20the%20future%20but%20are%20not%20used%20that%20much%20and%20are%20complex%20to%20implement%3B%20custom%20dtypes%20would%20increase%20design%20complexity%20that%20is%20not%20justified.) we should probably raise `NotImplementedError` when calling `col.dtype` in the case of object data type. `NotImplementedError` might also be raised when calling `col.describe_null` if `col.dtype` is called internally.

cc @vnlitvinov